### DISCUSSION TITLE
[Question] How to automatically fill android version info from AssemblyVersion

### DISCUSSION THREAD

**Discussion by janseris** [dotnet/maui#21427] [Q&A]
### Description

Hi,
Google Play and Android overall in order to correctly detect newer versions of applications, uses `<ApplicationVersion>` tagin .csproj which accepts an integer value (e.g. 5).
The standard for C#/.NET projects regarding versioning is <Version> which is Assembly Version (e.g. 1.2.3 or 1.2.3.4)
Then Android uses `<ApplicationDisplayVersion>` to display application version "name" to the users (e.g. "1.2.3")
How to please automatically synchronize these two from standard <Version>?For example, <ApplicationVersion> is required to be unique for every Android app Google Play store release.
Or at least how to synchronize assembly version (`<Version>`) with `<ApplicationDisplayVersion>` which uses the same value?

Thank you

**Comment by Ghostbird** [comment]
We never managed to do that either, so we chose a different approach. For the ApplicationDisplayVersion we use a semver string built from the git tag + commit count

For the ApplicationVersion we simply use the unix timestamp.

This can be done in the `csproj` file like this. Note that it has a dependency on `perl` in the PATH.

Also note that this does not prevent display name version conflicts from arising between branches.

```
<Project Sdk="Microsoft.NET.Sdk" InitialTargets="Set application display version; Set versions">
  <Target Name="Set application display version" Condition="'$(ApplicationDisplayVersion)' == '1.0.0'">
    <PropertyGroup>
      <Redirect Condition="Exists('/dev/null')">/dev/null</Redirect>
      <Redirect Condition="'$(Redirect)' == ''">NUL</Redirect>
    </PropertyGroup>
    <Exec Command="git describe --tags | perl -pe &quot;s/^v([0-9]+)(?:[.-]([0-9]+))?(?:[.-]([0-9]+))?(?:[.-]([0-9]+))?.*$/\1.\2.\3.\4/; s/\.+$//;&quot; 2&gt; $(Redirect)" ConsoleToMsBuild="True">
      <Output TaskParameter="ConsoleOutput" PropertyName="ApplicationDisplayVersion" />
    </Exec>
  </Target>
  <Target Name="Set versions" Condition="'$(ApplicationVersion)' == '1' AND '$(Configuration)' != 'Debug'">
    <Exec Command="date +%s" ConsoleToMsBuild="True">
      <Output TaskParameter="ConsoleOutput" PropertyName="ApplicationVersion" />
    </Exec>
  </Target>
 …
  ```
